app/Http/Requests/CloneVolume.php

```diff
@@ -0,0 +1,171 @@
+                        $validator->errors()->add('file_ids',
+                            'Unauthorized access to files that do not belong to the volume');
+                    }
+                } catch (Exception $e) {
+                    $validator->errors()->add('file_ids', $e->getMessage());
+                    return;
+                }
+
+                //check selected file label ids
+                try {
+                    // get file ids from label
+                    $fileLabelFileIds = $volume->isImageVolume() ?
+                        ImageLabel::findOrFail($fileLabelIds)->pluck('image_id')->toArray() :
+                        VideoLabel::findOrFail($fileLabelIds)->pluck('video_id')->toArray();
+                    $fileLabelFileIds = array_unique($fileLabelFileIds);
+
+                    // check file ids
+                    if (!empty($fileLabelFileIds) &&
+                        !$this->filesBelongToVolume($fileLabelFileIds, $volume->id, $volume->isImageVolume())) {
+                        $validator->errors()->add('$fileLabelFileIds',
+                            'Unauthorized access to labels that do not belong to the volume');
+                    }
+                } catch (Exception $e) {
+                    $validator->errors()->add('$fileLabelFileIds', $e->getMessage());
+                    return;
+                }
+
+                //check selected annotation label ids
+                try {
+                    // get annotation ids from annotations labels
+                    $fileAnnotationIds = $volume->isImageVolume() ?
+                        ImageAnnotationLabel::findOrFail($annotationLabelIds)->pluck('annotation_id')->toArray() :
+                        VideoAnnotationLabel::findOrFail($annotationLabelIds)->pluck('annotation_id')->toArray();
+                    $fileAnnotationIds = array_unique($fileAnnotationIds);
+
+                    // get file ids from annotations
+                    $annotationLabelFileIds = $volume->isImageVolume() ?
+                        ImageAnnotation::findOrFail($fileAnnotationIds)->pluck('image_id')->toArray() :
+                        VideoAnnotation::findOrFail($fileAnnotationIds)->pluck('video_id')->toArray();
+                    $annotationLabelFileIds = array_unique($annotationLabelFileIds);
+
+                    //check file ids
+                    if (!empty($annotationLabelFileIds) &&
+                        !$this->filesBelongToVolume($annotationLabelFileIds, $volume->id, $volume->isImageVolume())) {
+                        $validator->errors()->add('$fileLabelFileIds',
+                            'Unauthorized access to annotation labels that do not belong to the volume');
+                    }
+                } catch (Exception $e) {
+                    $validator->errors()->add('$fileLabelFileIds', $e->getMessage());
+                }
```

Comment:
There is a little misconception here. You expect IDs of the `image_labels`/`video_labels` tables in the `file_label_ids` parameter. Also, you expect IDs of the `image_annotation_labels`/`video_annotation_labels` tables in the `label_ids` parameter. But in both cases you should rather expect IDs of the `labels` table. Here are two examples to visualize this:

Filtering of image labels

```php
$labelIds = $this->input('file_label_ids', []);
// Get all image labels of a volume that a certain label_id
$imageLabels = ImageLabel::join('images', 'images.id', '=', 'image_annotations.image_id')
   ->where('images.volume_id', $volume->id)
   ->whereIn('label_id', $labelIds)
   ->get();
```

Filtering of annotations

```php
$labelIds = $this->input('annotation_label_ids', []);
// Get all annotations of a volume that have labels with a certain label_id
$annotations = ImageAnnotation::join('images', 'images.id', '=', 'image_annotations.image_id')
   ->where('images.volume_id', $volume->id)
   ->with('labels')
   ->whereHas('labels', function ($query) use ($labelIds) {
      $query->whereIn('label_id', $labelIds)
   })
   ->get();
```

(more on [`whereHas()`](https://laravel.com/docs/9.x/eloquent-relationships#querying-relationship-existence))

So there is no need to check if the ImageLabels/VideoLabels/ImageAnnotationLabels/VideoAnnotationLabels actually belong to the volume here. You just filter by the `label_id`. If there are no ImageLabels/VideoLabels/ImageAnnotationLabels/VideoAnnotationLabels with a particular label ID, none will be copied.

Reply:
So the whole withValidator function can be deleted?

Reply:
Not completely. It still makes sense to check if the `file_ids` all belong to the volume.